src/logger/mongoose_json_formatter.erl

```diff
@@ -0,0 +1,107 @@
+%%%-------------------------------------------------------------------
+%%% @copyright 2020, Erlang Solutions Ltd.
+%%% @doc
+%%% A module formatting logger reports into JSON objects.
+%%%
+%%% @end
+%%%-------------------------------------------------------------------
+-module(mongoose_json_formatter).
+
+%% API
+-export([format/2]).
+
+-spec format(logger:log_event(), logger:formatter_config()) -> unicode:chardata().
+format(E = #{msg := {report, #{label := {error_logger, _}, format := Format, args := Terms}}}, FConfig) ->
+    format(E#{msg := {report,
+                      #{unstructured_log =>
+                        unicode:characters_to_binary(io_lib:format(Format, Terms))}}},
+           FConfig);
+format(E = #{msg := {string, String}}, FConfig) ->
+    format(E#{msg := {report,
+                      #{unstructured_log =>
+                        unicode:characters_to_binary(io_lib:format(String, []))}}}, FConfig);
+format(E = #{msg := {report, Report}}, FConfig) when is_map(Report) ->
+    NewMap = process_metadata(E),
+    NewReport = maps:merge(Report, NewMap),
+    NewConfig = maps:merge(default_config(), config_correct_depth(FConfig)),
+    Formatted = format_item(NewReport, NewConfig),
+    unicode:characters_to_binary([jiffy:encode(Formatted), "\n"], utf8);
+format(Map = #{msg := {Format, Terms}}, FConfig) ->
+    format(Map#{msg := {report,
+                        #{unstructured_log =>
+                          unicode:characters_to_binary(io_lib:format(Format, Terms))}}},
+           FConfig).
+
+format_item(_Item, _FConfig = #{depth := 0}) ->
+    <<"...">>;
+format_item(Item, FConfig) when is_map(Item) ->
+    ML = [format_item({Key, Val}, FConfig) || {Key, Val} <- maps:to_list(Item)],
+    lists:foldl(fun(Map1, Acc) -> maps:merge(Map1, Acc) end, #{}, ML);
+format_item({K, V}, FConfig = #{depth := Depth}) ->
+    % Keys need to be strings in JSON
+    % We can get a nested structure as a key K here, it needs to be stringified
+    #{all_to_binary(K, FConfig) => format_item(V, FConfig#{depth := Depth - 1})};
+format_item(Item, FConfig = #{depth := Depth}) when is_list(Item) ->
+    case io_lib:printable_unicode_list(Item) of
+        true ->
+            % We want to print strings as strings
+            format_str(Item, FConfig);
```

Comment:
Do we REALLY use any unicode character **lists** in mongooseim?
All unicode strings are binaries.

We should print lists as lists I think.

Reply:
There are some, maybe not from MIM, but for example the `file` in the logger event metadata is a string.